File: ragemp-server/packages/core/elections.js
// Sistem alegeri Presedinte
// Comanda: /euvotez [nume_jucator]
// Alegerile au loc in fiecare duminica (la sfarsit se desemneaza castigatorul cu cele mai multe voturi)
// Presedintele controleaza suma1 (bonus salariu factiune, 0-5000) si suma2 (bonus CASS, 0-1000)

const cron = require('node-cron');
const db = require('./database');

const SUMA1_MIN = 0, SUMA1_MAX = 5000;
const SUMA2_MIN = 0, SUMA2_MAX = 1000;

async function getActiveElection() {
    const now = new Date();
    return await db.one(
        'SELECT * FROM elections WHERE start_date <= ? AND end_date >= ? AND finalizata = 0 ORDER BY id DESC LIMIT 1',
        [now, now]
    );
}

async function createElectionForCurrentSunday() {
    // Creeaza alegere pentru duminica curenta (00:00 - 23:59 ora serverului)
    const now = new Date();
    const day = now.getDay(); // 0 = duminica
    const start = new Date(now);
    start.setHours(0, 0, 0, 0);
    if (day !== 0) {
        // sari la urmatoarea duminica
        start.setDate(start.getDate() + (7 - day));
    }
    const end = new Date(start);
    end.setHours(23, 59, 59, 999);

    const existing = await db.one(
        'SELECT id FROM elections WHERE start_date = ?',
        [start]
    );
    if (existing) return existing.id;

    const res = await db.query('INSERT INTO elections (start_date, end_date) VALUES (?, ?)', [start, end]);
    console.log(`[ALEGERI] Alegere creata pentru duminica ${start.toISOString()}`);
    return res.insertId;
}

async function vote(votantDbId, candidatNume) {
    const election = await getActiveElection();
    if (!election) return { ok: false, msg: 'Nu exista alegeri active. Voturile se accepta doar duminica.' };

    const candidat = await db.one('SELECT id, nume FROM players WHERE nume = ?', [candidatNume]);
    if (!candidat) return { ok: false, msg: `Jucatorul "${candidatNume}" nu exista.` };

    try {
        await db.query(
            'INSERT INTO votes (election_id, votant_id, candidat_id) VALUES (?, ?, ?)',
            [election.id, votantDbId, candidat.id]
        );
        return { ok: true, msg: `Ai votat pentru ${candidat.nume}.` };
    } catch (e) {
        if (e.code === 'ER_DUP_ENTRY') return { ok: false, msg: 'Ai votat deja la aceasta alegere.' };
        throw e;
    }
}

async function tally(electionId) {
    return await db.query(`
        SELECT p.id, p.nume, COUNT(v.id) AS voturi
        FROM votes v
        JOIN players p ON p.id = v.candidat_id
        WHERE v.election_id = ?
        GROUP BY p.id, p.nume
        ORDER BY voturi DESC, p.nume ASC
    `, [electionId]);
}

async function finalizeElection(electionId) {
    const results = await tally(electionId);
    if (results.length === 0) {
        await db.query('UPDATE elections SET finalizata = 1 WHERE id = ?', [electionId]);
        console.log(`[ALEGERI] Alegerea #${electionId} finalizata fara voturi.`);
        return null;
    }
    const winner = results[0];
    await db.query('UPDATE elections SET finalizata = 1, castigator_id = ? WHERE id = ?', [winner.id, electionId]);
    await db.query('UPDATE president SET player_id = ?, inceput_mandat = NOW() WHERE id = 1', [winner.id]);
    console.log(`[ALEGERI] Castigator: ${winner.nume} (${winner.voturi} voturi)`);
    return winner;
}

async function setPresidentTaxes(playerDbId, suma1, suma2) {
    const pres = await db.one('SELECT player_id FROM president WHERE id = 1');
    if (!pres || pres.player_id !== playerDbId) {
        return { ok: false, msg: 'Doar Presedintele poate seta taxele.' };
    }
    if (suma1 < SUMA1_MIN || suma1 > SUMA1_MAX) return { ok: false, msg: `Suma1 trebuie intre ${SUMA1_MIN}-${SUMA1_MAX}.` };
    if (suma2 < SUMA2_MIN || suma2 > SUMA2_MAX) return { ok: false, msg: `Suma2 trebuie intre ${SUMA2_MIN}-${SUMA2_MAX}.` };

    await db.query('UPDATE president SET suma1 = ?, suma2 = ? WHERE id = 1', [suma1, suma2]);
    return { ok: true, msg: `Taxele actualizate: suma1=${suma1}, suma2=${suma2}` };
}

async function getCurrentPresident() {
    return await db.one(`
        SELECT pr.suma1, pr.suma2, pr.inceput_mandat, p.id AS player_id, p.nume
        FROM president pr LEFT JOIN players p ON p.id = pr.player_id
        WHERE pr.id = 1
    `);
}

function start() {
    // Asigura alegere creata pentru duminica curenta la pornire si in fiecare zi la 00:01
    cron.schedule('1 0 * * *', async () => {
        try { await createElectionForCurrentSunday(); } catch (e) { console.error('[ALEGERI] eroare cron creare', e); }
    });
    // Finalizeaza alegerile duminica la 23:55
    cron.schedule('55 23 * * 0', async () => {
        try {
            const el = await getActiveElection();
            if (el) await finalizeElection(el.id);
        } catch (e) { console.error('[ALEGERI] eroare finalizare', e); }
    });
    createElectionForCurrentSunday().catch(e => console.error(e));
    console.log('[ALEGERI] Scheduler pornit.');
}

module.exports = {
    start, vote, tally, finalizeElection,
    setPresidentTaxes, getCurrentPresident,
    createElectionForCurrentSunday, getActiveElection,
    SUMA1_MIN, SUMA1_MAX, SUMA2_MIN, SUMA2_MAX
};

File: fivem-server/resources/[rrp]/rrp-core/server/elections.lua
-- ============================================================
--  ALEGERI PRESEDINTE - voturile doar duminica
--  /euvotez <nume>          - 1 vot/saptamana/jucator
--  /presedinte              - vezi presedintele si taxele curente
--  /setezetaxe <s1> <s2>    - doar Presedinte. s1[0-5000] s2[0-1000]
--  /finalizeazaalegerile    - doar admin lvl 10
-- ============================================================

local function getCurrentSundayBoundary()
    -- Returneaza inceputul (00:00) si sfarsitul (23:59:59) duminicii curente (sau urmatoare)
    local now = os.time()
    local t = os.date('*t', now)
    -- wday: 1=Sunday, 7=Saturday in Lua os.date
    local daysToSunday = (1 - t.wday) % 7
    if daysToSunday == 0 and (t.hour > 23 or (t.hour == 23 and t.min >= 55)) then
        daysToSunday = 7
    end
    local startTs = now + daysToSunday * 86400
    local st = os.date('*t', startTs)
    st.hour, st.min, st.sec = 0, 0, 0
    local startEpoch = os.time(st)
    local endEpoch = startEpoch + 86399
    return startEpoch, endEpoch
end

local function ensureActiveElection()
    local startE, endE = getCurrentSundayBoundary()
    local startStr = os.date('%Y-%m-%d %H:%M:%S', startE)
    local endStr = os.date('%Y-%m-%d %H:%M:%S', endE)
    local existing = MySQL.single.await('SELECT id FROM rrp_elections WHERE start_date = ?', { startStr })
    if existing then return existing.id, startE, endE end
    local id = MySQL.insert.await('INSERT INTO rrp_elections (start_date, end_date) VALUES (?, ?)', { startStr, endStr })
    return id, startE, endE
end

local function isVotingDay()
    -- Permite doar duminica intre 00:00 si 23:59
    return os.date('*t').wday == 1
end

QBCore.Commands.Add(Config.Elections.voteCommand, 'Voteaza Presedinte',
    { { name = 'nume', help = 'Numele candidatului (prenume/familie)' } }, true, function(source, args)
        if not isVotingDay() then
            return TriggerClientEvent('QBCore:Notify', source, 'Voturile se accepta doar duminica.', 'error')
        end
        local voterCid = RRP_GetCitizenId(source)
        if not voterCid then return end
        local nume = table.concat(args, ' ')

        -- candidat: cauta online sau in db players (qb-core characters)
        local candCid, candNume
        local _, candPlayer = RRP_FindPlayerByName(nume)
        if candPlayer then
            candCid = candPlayer.PlayerData.citizenid
            candNume = (candPlayer.PlayerData.charinfo.firstname or '') .. ' ' .. (candPlayer.PlayerData.charinfo.lastname or '')
        else
            -- cauta offline in db
            local row = MySQL.single.await([[
                SELECT citizenid, JSON_VALUE(charinfo, '$.firstname') AS firstname, JSON_VALUE(charinfo, '$.lastname') AS lastname
                FROM players
                WHERE LOWER(CONCAT(JSON_VALUE(charinfo, '$.firstname'), ' ', JSON_VALUE(charinfo, '$.lastname'))) LIKE LOWER(?)
                LIMIT 1
            ]], { '%'..nume..'%' })
            if row then candCid = row.citizenid; candNume = row.firstname..' '..row.lastname end
        end
        if not candCid then return TriggerClientEvent('QBCore:Notify', source, 'Candidat inexistent: '..nume, 'error') end

        local electionId = ensureActiveElection()
        local already = MySQL.scalar.await('SELECT id FROM rrp_votes WHERE election_id = ? AND votant_cid = ?', { electionId, voterCid })
        if already then return TriggerClientEvent('QBCore:Notify', source, 'Ai votat deja la aceasta alegere.', 'error') end

        MySQL.insert('INSERT INTO rrp_votes (election_id, votant_cid, candidat_cid, candidat_nume) VALUES (?, ?, ?, ?)',
            { electionId, voterCid, candCid, candNume })
        TriggerClientEvent('QBCore:Notify', source, 'Vot inregistrat pentru '..candNume, 'success')
    end)

QBCore.Commands.Add('presedinte', 'Vezi Presedintele curent', {}, false, function(source)
    local row = MySQL.single.await('SELECT * FROM rrp_president WHERE id = 1')
    if not row or not row.citizenid then
        return TriggerClientEvent('QBCore:Notify', source, 'Nu exista Presedinte ales.', 'primary')
    end
    TriggerClientEvent('QBCore:Notify', source, string.format('Presedinte: %s | suma1=%d suma2=%d', row.nume or '?', row.suma1, row.suma2), 'primary')
end)

QBCore.Commands.Add('setezetaxe', 'Presedinte: seteaza suma1 si suma2',
    { { name = 'suma1', help = '0-5000' }, { name = 'suma2', help = '0-1000' } }, true, function(source, args)
        local cid = RRP_GetCitizenId(source)
        local pres = MySQL.single.await('SELECT citizenid FROM rrp_president WHERE id = 1')
        if not pres or pres.citizenid ~= cid then
            return TriggerClientEvent('QBCore:Notify', source, 'Doar Presedintele poate.', 'error')
        end
        local s1, s2 = tonumber(args[1]), tonumber(args[2])
        if not s1 or s1 < Config.Elections.suma1Min or s1 > Config.Elections.suma1Max then
            return TriggerClientEvent('QBCore:Notify', source, 'suma1 trebuie ['..Config.Elections.suma1Min..'-'..Config.Elections.suma1Max..']', 'error')
        end
        if not s2 or s2 < Config.Elections.suma2Min or s2 > Config.Elections.suma2Max then
            return TriggerClientEvent('QBCore:Notify', source, 'suma2 trebuie ['..Config.Elections.suma2Min..'-'..Config.Elections.suma2Max..']', 'error')
        end
        MySQL.update('UPDATE rrp_president SET suma1 = ?, suma2 = ? WHERE id = 1', { s1, s2 })
        TriggerClientEvent('QBCore:Notify', source, 'Taxe actualizate: s1='..s1..' s2='..s2, 'success')
    end)

QBCore.Commands.Add('finalizeazaalegerile', 'Admin 10: finalizeaza alegerea', {}, false, function(source)
    if RRP_GetAdminLevel(source) < 10 then
        return TriggerClientEvent('QBCore:Notify', source, '[Admin] Lvl 10 necesar.', 'error')
    end
    local electionId = ensureActiveElection()
    local rows = MySQL.query.await('SELECT candidat_cid, candidat_nume, COUNT(*) AS voturi FROM rrp_votes WHERE election_id = ? GROUP BY candidat_cid, candidat_nume ORDER BY voturi DESC LIMIT 1', { electionId })
    if not rows or #rows == 0 then
        MySQL.update('UPDATE rrp_elections SET finalizata = 1 WHERE id = ?', { electionId })
        return TriggerClientEvent('QBCore:Notify', source, 'Alegere finalizata fara voturi.', 'primary')
    end
    local w = rows[1]
    MySQL.update('UPDATE rrp_elections SET finalizata = 1, castigator_cid = ? WHERE id = ?', { w.candidat_cid, electionId })
    MySQL.update('UPDATE rrp_president SET citizenid = ?, nume = ?, inceput_mandat = NOW() WHERE id = 1', { w.candidat_cid, w.candidat_nume })
    TriggerClientEvent('QBCore:Notify', source, 'Castigator: '..w.candidat_nume..' ('..w.voturi..' voturi)', 'success')
end)
